File: WebSockets-Joust/src/java/model/Joust.java
package model;

import java.util.Arrays;
import java.util.stream.Stream;

public class Joust {

    private final int rows;
    private final int cols;
    private Player turn = Player.PLAYER1;
    private final CellState[][] board;

    public Joust(int nrows, int ncols) {
        this.rows = nrows;
        this.cols = ncols;
        this.board = this.startBoard();
    }

    private CellState[][] startBoard() {
        CellState[][] matrix = new CellState[rows][cols];
        for (CellState[] row : matrix) {
            Arrays.fill(row, CellState.EMPTY);
        }
        CellState[] players = {CellState.PLAYER1, CellState.PLAYER2};
        for (int i = 0; i < players.length;) {
            int row = (int) Math.floor(Math.random() * rows);
            int col = (int) Math.floor(Math.random() * cols);
            if (matrix[row][col] == CellState.EMPTY) {
                matrix[row][col] = players[i];
                i++;
            }
        }
        return matrix;
    }

    public Move move(Player player, Cell endCell) {
        if (player != turn) {
            return Move.INVALID;
        }
        Cell beginCell = getPlayerCell(player);
        Move m = isValidMove(beginCell, endCell);
        if (m == Move.VALID) {
            int or = beginCell.getX(), oc = beginCell.getY();
            int dr = endCell.getX(), dc = endCell.getY();
            board[dr][dc] = board[or][oc];
            board[or][oc] = CellState.BLOCKED;
            turn = (turn == Player.PLAYER1) ? Player.PLAYER2 : Player.PLAYER1;
        }
        return m;
    }

    private Move isValidMove(Cell beginCell, Cell endCell) {
        int or = beginCell.getX(), oc = beginCell.getY();
        int dr = endCell.getX(), dc = endCell.getY();
        /* Destino deve estar vazio */
        if (board[dr][dc] != CellState.EMPTY) {
            return Move.INVALID;
        }
        int[] lin = {-2, 2, -1, 1};
        int[][] inc = {{-1, 1}, {-1, 1}, {-2, 2}, {-2, 2}};
        for (int i = 0; i < lin.length; i++) {
            for (int j = 0; j < inc[i].length; j++) {
                Cell cell = new Cell(or + lin[i], oc + inc[i][j]);
                if (isValidCell(cell) && cell.equals(endCell)) {
                    return Move.VALID;
                }
            }
        }
        return Move.INVALID;
    }

    private Cell getPlayerCell(Player player) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if ((board[i][j] == CellState.PLAYER1 && player == Player.PLAYER1)
                        || (board[i][j] == CellState.PLAYER2 && player == Player.PLAYER2)) {
                    return new Cell(i, j);
                }
            }
        }
        return null;
    }

    private boolean isValidCell(Cell cell) {
        return (cell.getX() < rows && cell.getX() >= 0 && cell.getY() < cols && cell.getY() >= 0);
    }

    public Player getTurn() {
        return turn;
    }

    public CellState[][] getBoard() {
        return board;
    }

    public static void main(String[] args) {
        Joust j = new Joust(8, 8);
        Stream.of(j.getBoard()).forEach(x -> System.out.println(Arrays.toString(x)));
    }
}

File: docs/scripts/Joust.js
import CellState from './CellState.js';
import Player from './Player.js';
import Cell from './Cell.js';
import Winner from './Winner.js';

export default class Joust {
    constructor(nrows, ncols) {
        this.rows = nrows;
        this.cols = ncols;
        this.turn = Player.PLAYER1;    
        this.board = this.startBoard();
        this.winner = Winner.NONE;
    }
    getWinner() {
        return this.winner;
    }
    getBoard() {
        return this.board;
    }
    startBoard() {
        let matrix = Array(this.rows).fill().map(() => Array(this.cols).fill(CellState.EMPTY));
        let players = [CellState.PLAYER1, CellState.PLAYER2];
        for (let i = 0; i < players.length;) {
            let row = Math.floor(Math.random() * this.rows);
            let col = Math.floor(Math.random() * this.cols);
            if (matrix[row][col] === CellState.EMPTY) {
                matrix[row][col] = players[i++];
            }
        }
        return matrix;
    }
    getTurn() {
        return this.turn;
    }
    move(beginCell, endCell) {
        if(this.winner !== Winner.NONE) {
            throw new Error("This game is already finished.");
        }
        let { x: or, y: oc } = beginCell;
        let { x: dr, y: dc } = endCell;
        if (!beginCell || !endCell) {
            throw new Error("The value of one of the cells does not exist.");
        }
        let currentPiece = this.getPiece(beginCell);
        if ((currentPiece === CellState.PLAYER1 && this.turn === Player.PLAYER2) || (currentPiece === CellState.PLAYER2 && this.turn === Player.PLAYER1)) {
            throw new Error("It's not your turn.");
        }
        if (beginCell.equals(endCell)) {
            throw new Error("Origin and destination must be different.");
        }
        if (!this.onBoard(beginCell) || !this.onBoard(endCell)) {
            throw new Error("Origin or destination are not in the board.");
        }
        if (currentPiece === CellState.EMPTY) {
            throw new Error("Origin does not have a piece.");
        }
        if (this.getPiece(endCell) !== CellState.EMPTY) {
            throw new Error("Destination must be empty.");
        }
        let moves = this.possibleMoves(beginCell);
        if (!moves.some(z => z.equals(endCell))) {
            throw new Error("This move is invalid.");
        }
        this.board[dr][dc] = this.board[or][oc];
        this.board[or][oc] = CellState.BLOCKED;
        this.turn = (this.turn === Player.PLAYER1) ? Player.PLAYER2 : Player.PLAYER1;
        this.winner = this.endOfGame();
    }
    getPiece({ x, y }) {
        return this.board[x][y];
    }
    onBoard({ x, y }) {
        let inLimit = (value, limit) => value >= 0 && value < limit;
        return (inLimit(x, this.rows) && inLimit(y, this.cols));
    }
    possibleMoves({ x, y }, matrix = this.board) {
        let positions = [new Cell(x - 2, y - 1), new Cell(x - 2, y + 1), new Cell(x + 2, y - 1), new Cell(x + 2, y + 1), new Cell(x - 1, y - 2), new Cell(x - 1, y + 2), new Cell(x + 1, y - 2), new Cell(x + 1, y + 2)];
        return positions.filter(cell => this.onBoard(cell) && matrix[cell.x][cell.y] === CellState.EMPTY);
    }
    endOfGame() {
        let p1 = this.canMove(Player.PLAYER1);
        let p2 = this.canMove(Player.PLAYER2);
        if (!p1 && !p2) {
            return Winner.DRAW;
        } else if (!p1) {
            return Winner.PLAYER2;
        } else if (!p2) {
            return Winner.PLAYER1;
        }
        return Winner.NONE;
    }
    canMove(player) {
        let i = this.board.flat().indexOf(player === Player.PLAYER1 ? CellState.PLAYER1 : CellState.PLAYER2);
        let cell = new Cell(Math.floor(i / this.cols), i % this.cols);
        return this.board.flat().some((e, i) => this.isValidMove(cell, new Cell(Math.floor(i / this.cols), i % this.cols)));
    }
    isValidMove(player, endCell) {
        let moves = this.possibleMoves(player);
        return moves.some(e => e.equals(endCell));
    }
}